File: ex2-linear-regression/ex2.m
x = load('ex2x.dat');
y = load('ex2y.dat');

figure; % open a new figure window
plot(x, y, 'o');
ylabel('Height in meters')
xlabel('Age in years')

m = length(y); % store the number of training examples
x = [ones(m, 1), x]; % Add a column of ones to x

alpha = 0.07;
theta = [0; 0];

delta = - alpha * (1/m) * x' * (x * theta - y);
theta = theta + delta;

theta1 = theta;

while (norm(delta) > 1e-6)
  delta = - alpha * (1/m) * x' * (x * theta - y);
  theta = theta + delta;
end

hold on; % Plot new data without clearing old plot
plot(x(:,2), x*theta, '-'); % remember that x is now a matrix with 2 columns
                            % and the second column contains the time info
legend('Training data', 'Linear regression');


J_vals = zeros(100, 100);   % initialize Jvals to 100x100 matrix of 0's
theta0_vals = linspace(-3, 3, 100);
theta1_vals = linspace(-1, 1, 100);
for i = 1:length(theta0_vals)
  for j = 1:length(theta1_vals)
    t = [theta0_vals(i); theta1_vals(j)];
    J_vals(i,j) = (1 / (2 * m)) * sum((x * t - y) .^ 2);
  end
end

% Plot the surface plot
% Because of the way meshgrids work in the surf command, we need to
% transpose J_vals before calling surf, or else the axes will be flipped
J_vals = J_vals';
figure;
surf(theta0_vals, theta1_vals, J_vals);
xlabel('\theta_0'); ylabel('\theta_1');

figure;
% Plot the cost function with 15 contours spaced logarithmically
% between 0.01 and 100
contour(theta0_vals, theta1_vals, J_vals, logspace(-2, 2, 15));
xlabel('\theta_0'); ylabel('\theta_1');



% print out values of interest (no semicolons)

theta1 = theta1
theta = theta

ages = [ 1, 3.5; 1, 7 ];
heights = ages * theta

File: ex2-linear-regression/ex2.py
#! ~/.ve/np/bin/python

import time
import sys
import numpy as np
import scipy as sp
import matplotlib as mpl
if sys.platform == 'darwin':
  mpl.use('TkAgg')
import matplotlib.pyplot as plt
from scipy import linalg
from mpl_toolkits.mplot3d import Axes3D

X = np.loadtxt('ex2x.dat')[:, np.newaxis]
y = np.loadtxt('ex2y.dat')[:, np.newaxis]

plt.figure()
plt.plot(X, y, 'o')
plt.ylabel('Height in meters')
plt.xlabel('Age in years')

X = np.hstack([np.ones([X.shape[0], 1]), X])
m, n = X.shape

alpha = 0.07
theta = np.zeros([n, 1])

delta = - alpha * (1.0/m) * X.T.dot(X.dot(theta) - y)
theta += delta

print theta

while linalg.norm(delta) > 1e-6:
  delta = - alpha * (1.0/m) * X.T.dot(X.dot(theta) - y)
  theta += delta

print theta

plt.plot(X[:, 1:], X.dot(theta), '-')
# plt.legend('Training data', 'Linear regression')

J_vals = np.zeros([100, 100])
theta0_vals = np.linspace(-3, 3, 100)
theta1_vals = np.linspace(-1, 1, 100)

for i in range(0, theta0_vals.size):
  for j in range(0, theta1_vals.size):
      t = np.array([[theta0_vals[i]], [theta1_vals[j]]])
      J_vals[i, j] = (1.0 / (2 * m)) * np.sum( (X.dot(t) - y) ** 2)

J_vals = J_vals.T

fig = plt.figure()
ax = fig.gca(projection='3d')
xx, yy = np.meshgrid(theta0_vals, theta1_vals)

surf = ax.plot_surface(xx, yy, J_vals, rstride=1, cstride=1, cmap=mpl.cm.jet,
                       linewidth=0, antialiased=False)



plt.figure()
plt.contour(theta0_vals, theta1_vals, J_vals, np.logspace(-2, 2, 15));





plt.show()
